File: app/src/main/java/farasense/mobile/service/base/BaseService.java
package farasense.mobile.service.base;

import android.app.Service;
import android.content.Intent;
import android.os.Handler;
import android.os.IBinder;
import android.os.Looper;
import android.os.Message;
import android.support.annotation.Nullable;
import android.util.Log;

import farasense.mobile.service.download.DownloadFaraSenseSensorDailyService;
import farasense.mobile.service.download.DownloadFaraSenseSensorHoursService;
import farasense.mobile.service.download.DownloadFaraSenseSensorService;
import farasense.mobile.service.listener.OnDownloadContentListener;
import farasense.mobile.service.listener.OnStartServiceDownload;
import farasense.mobile.util.DateUtil;

public class BaseService extends Service {

    public static final int totalInitialsServices = 3;
    protected static final int SLEEP_WAIT = 300000;
    protected static final int SLEEP_TRY_AGAIN = 5000;
    protected static final int SLEEP_TRY_AGAIN_DOUBLE = 10000;
    protected static final String LOG_DSERVICE_OK = "DOWNLOAD ;) Serviço realizado com sucesso!";
    protected static final String LOG_DSERVICE_ERROR = "DOWNLOAD :( Serviço falhou...";
    protected static final String LOG_USERVICE_OK = "UPLOAD ;) Serviço realizado com sucesso!";
    protected static final String LOG_USERVICE_ERROR = "UPLOAD :( Serviço falhou...";

    protected Looper looper;
    protected Message msg;
    protected boolean shouldStop;
    protected Handler handler;

    @Override
    public void onCreate() {
        shouldStop = false;
    }

    @Nullable
    @Override
    public IBinder onBind(Intent intent) {
        return null;
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        stopSelf(msg.arg1);
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        msg = handler.obtainMessage();
        msg.arg1 = startId;
        handler.sendMessage(msg);
        return START_STICKY;
    }

    public static void initialDownloadDataSensors(final OnStartServiceDownload onStartServiceDownload) {
        final int[] finishedServices = {0};
        onStartServiceDownload.onStart();

        DownloadFaraSenseSensorService.download(new OnDownloadContentListener() {
            @Override
            public void onSucess() {
                finishedServices[0]++;

                if (finishedServices[0] >= totalInitialsServices) {
                    onStartServiceDownload.onFinish();
                }
            }

            @Override
            public void onFail() {
                onStartServiceDownload.onFail();

            }
        }, DateUtil.getFirts24Hours(), DateUtil.getNow());

        DownloadFaraSenseSensorHoursService.download(new OnDownloadContentListener() {
            @Override
            public void onSucess() {
                finishedServices[0]++;

                if (finishedServices[0] >= totalInitialsServices) {
                    onStartServiceDownload.onFinish();
                }
            }

            @Override
            public void onFail() {
                onStartServiceDownload.onFail();
            }
        }, DateUtil.getFirts24HoursWithMinutesReset(), DateUtil.getNow());

        DownloadFaraSenseSensorDailyService.download(new OnDownloadContentListener() {
            @Override
            public void onSucess() {
                finishedServices[0]++;

                if (finishedServices[0] >= totalInitialsServices) {
                    onStartServiceDownload.onFinish();
                }
            }

            @Override
            public void onFail() {
                onStartServiceDownload.onFail();;
            }
        }, DateUtil.getFirts30Days(), DateUtil.getNow());

    }
}

File: app/src/main/java/farasense/mobile/service/base/BaseService.kt
package farasense.mobile.service.base

import android.annotation.SuppressLint
import android.app.Service
import android.content.Context
import android.content.Intent
import android.os.Handler
import android.os.IBinder
import android.os.Looper
import android.os.Message
import farasense.mobile.service.download.DownloadFaraSenseSensorDailyService
import farasense.mobile.service.download.DownloadFaraSenseSensorHoursService
import farasense.mobile.service.download.DownloadFaraSenseSensorService
import farasense.mobile.service.listener.OnDownloadContentListener
import farasense.mobile.service.listener.OnStartServiceDownload
import farasense.mobile.util.DateUtil

@SuppressLint("Registered")
open class BaseService : Service() {

    val totalInitialsServices = 3

    protected var looper: Looper? = null
    protected lateinit var msg: Message
    protected var shouldStop: Boolean = false
    protected var handler: Handler? = null
    protected var context: Context? = null

    override fun onCreate() {
        shouldStop = false
        context = applicationContext
    }

    override fun onBind(intent: Intent): IBinder? {
        return null
    }

    override fun onDestroy() {
        super.onDestroy()
        stopSelf(msg.arg1)
    }

    override fun onStartCommand(intent: Intent, flags: Int, startId: Int): Int {
        msg = handler!!.obtainMessage()
        msg.arg1 = startId
        handler!!.sendMessage(msg)
        return START_STICKY
    }

    fun initialDownloadDataSensors(context: Context, onStartServiceDownload: OnStartServiceDownload) {
        val finishedServices = intArrayOf(0)
        onStartServiceDownload.onStart()

        DownloadFaraSenseSensorService.download(context, object : OnDownloadContentListener {
            override fun onSucess() {
                finishedServices[0]++

                if (finishedServices[0] >= totalInitialsServices) {
                    onStartServiceDownload.onFinish()
                }
            }

            override fun onFail() {
                onStartServiceDownload.onFail()
            }
        }, DateUtil.firts24Hours, DateUtil.now)

        DownloadFaraSenseSensorHoursService.download(context, object : OnDownloadContentListener {
            override fun onSucess() {
                finishedServices[0]++

                if (finishedServices[0] >= totalInitialsServices) {
                    onStartServiceDownload.onFinish()
                }
            }

            override fun onFail() {
                onStartServiceDownload.onFail()
            }
        }, DateUtil.firts24HoursWithMinutesReset, DateUtil.now)

        DownloadFaraSenseSensorDailyService.download(context, object : OnDownloadContentListener {
            override fun onSucess() {
                finishedServices[0]++

                if (finishedServices[0] >= totalInitialsServices) {
                    onStartServiceDownload.onFinish()
                }
            }

            override fun onFail() {
                onStartServiceDownload.onFail()
            }
        }, DateUtil.firts30Days, DateUtil.now)
    }

    companion object {
        val SLEEP_WAIT = 300000
        val SLEEP_TRY_AGAIN = 5000
        val SLEEP_TRY_AGAIN_DOUBLE = 10000
        val LOG_DSERVICE_OK = "DOWNLOAD ;) Serviço realizado com sucesso!"
        val LOG_DSERVICE_ERROR = "DOWNLOAD :( Serviço falhou..."
        val LOG_USERVICE_OK = "UPLOAD ;) Serviço realizado com sucesso!"
        val LOG_USERVICE_ERROR = "UPLOAD :( Serviço falhou..."
    }
}
